library/ssl_tls13_client.c

```diff
@@ -25,14 +25,455 @@
 
 #if defined(MBEDTLS_SSL_CLI_C)
 
+#include <string.h>
+
 #include "ssl_misc.h"
+#include <mbedtls/debug.h>
+
+/* Main entry point; orchestrates the other functions */
+static int ssl_client_hello_process( mbedtls_ssl_context* ssl );
 
 int mbedtls_ssl_handshake_client_step_tls1_3( mbedtls_ssl_context *ssl )
+{
+    int ret = 0;
+
+    if( ssl->state == MBEDTLS_SSL_HANDSHAKE_OVER || ssl->handshake == NULL )
+    {
+        MBEDTLS_SSL_DEBUG_MSG( 2, ( "Handshake completed but ssl->handshake is NULL.\n" ) );
+        return( MBEDTLS_ERR_SSL_BAD_INPUT_DATA );
+    }
+
+    MBEDTLS_SSL_DEBUG_MSG( 2, ( "client state: %d", ssl->state ) );
+
+    switch( ssl->state )
+    {
+        case MBEDTLS_SSL_HELLO_REQUEST:
```

Comment:
This is from before I joined the project, so I'm really not sure what its purpose initially was.

I _think_ nowadays it's used a special state that indicates that the context is fresh and the handshake hasn't started yet (see its use in `mbedtls_ssl_context_load() for example). Nowadays `MBEDTLS_SSL_CLIENT_HELLO` is ambiguous client-side with DTLS: if can mean either that the handshake hasn't started yet, or that we've just finished processing a HelloVerifyRequest. Note though, that before DTLS was implemented, `MBEDTLS_SSL_CLIENT_HELLO` was unambiguous and always meant that the handshake hadn't started - so at this point `MBEDTLS_SSL_CLIENT_HELLO` could have been used as the initial state AFAIU and I have no idea why `HELLO_REQUEST` was used instead. (Note that `HELLO_REQUEST` on its own is ambiguous as well: it means either that the context is fresh, or that renegotation is about to begin. Also, it's unclear to me whether ability to check if a context is fresh is really something we need.)

A wild guess would be that it was initially meant to serve some purposed related to server-initiated renegotiation - which, when I joined the project, had been broken forever (`ssl-opt.sh` didn't exist back then so most SSL/TLS features beyond a normal handshake were entirely untested). We'll probably never now. What's clear is that nowadays it doesn't serve any special purpose related to server-initiated renego, as we have a special field for that.

To be honest, it's not clear to me why this field was introduced, and whether it (still) serves a purpose now. I guess it was just kept because I was never 100% sure I wasn't missing something. Which is a sad state of affairs and only reinforces Ronald's point: we should really document what each state means. (For this state, probably document as "unclear" now and create a tech debt ticket to come back to it later.)

Reply:
I agree that this needs to be improved - I am only generally wary that we don't impede the tightly-scheduled 1.3 development by holding it to a higher standard than the 1.2 code.

Reply:
Yes, I agree, there's a fine balance to be found here. On one hand, I think it's easier to try to get the 1.3 code right on the first time than tracking improvements to be done later - and experience has also proven that it can very hard to get those improvements scheduled. OTOH, obviously we don't want to impede 1.3 development.

In that case though, I think since we're modifying an existing undocumented convention (that most states are named after the handshake message that will be expected or written next), we should add a small comment on the definition of the enum. I'll suggest one in the appropriate section as I realise this thread is not where the enum is defined.

Regarding this code, I think it should be kept as is this is the initial state of a freshly setup or configured context, so we need to handle it gracefully, by moving to the next state without doing anything else, as is done in TLS 1.2. Later we can come back and remove that dummy state for both 1.2 and 1.3 when we have the time.